servers/src/main/java/tachyon/master/next/IndexedSet.java

```diff
@@ -0,0 +1,231 @@
+/*
+ * Licensed to the University of California, Berkeley under one or more contributor license
+ * agreements. See the NOTICE file distributed with this work for additional information regarding
+ * copyright ownership. The ASF licenses this file to You under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package tachyon.master.next;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.HashMap;
+import java.util.Set;
+import java.util.HashSet;
+import java.lang.reflect.Field;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Sets;
+
+import tachyon.Constants;
+
+/**
+ * A set of objects that can be indexed by specific fields of the object, and different IndexedSet
+ * instances can use different fields to index. The field type must be comparable. The field value
+ * must not be changed after being added to the set, otherwise, behavior for all operations is not
+ * specified.
+ *
+ * In operations that need field name as the parameter, if the object does not have the field, a
+ * RuntimeException will be thrown.
+ *
+ * This class is thread safe.
+ */
+public class IndexedSet<T> {
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+
+  /** All objects in the set */
+  private final Set<T> mObjects;
+  /** Map from field index to an index of field related object in the internal lists */
+  private final Map<FieldIndex<T>, Integer> mIndexMap;
+  /** List of maps from fieldValue to set of T which is indexed by field name */
+  private final List<Map<Object, Set<T>>> mSetIndexedByFieldValue;
+  /** Final object for synchronization */
+  private final Object mLock = new Object();
+
+  /**
+   * A wrapper around a field name string to force the user to predefine the indexes, the user can
+   * reuse the instance of this class as parameters later other than directly use the error prone
+   * raw Strings.
+   */
+  public static interface FieldIndex<T> {
+    Object getFieldValue(T o);
+  }
+
+  /**
+   * Construct a new IndexedSet with at least one field as the index, the field can be either public
```

Comment:
I think this comment is incorrect now. FieldIndex instances are passed in, so the private/public comment doesn't seem correct anymore. Also, later sentences reference "Field", and reflection, which are not accurate anymore.